File: PHP/Endpoint Examples/JSON Payload/pdf-with-added-attachment.php
<?php
require 'vendor/autoload.php'; // Require the autoload file to load Guzzle HTTP client.

use GuzzleHttp\Client; // Import the Guzzle HTTP client namespace.
use GuzzleHttp\Psr7\Request; // Import the PSR-7 Request class.
use GuzzleHttp\Psr7\Utils; // Import the PSR-7 Utils class for working with streams.

$upload_pdf_client = new Client(['http_errors' => false]);
$upload_pdf_headers = [
  'api-key' => 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx',
  'content-filename' => 'pdf_filename.pdf',
  'Content-Type' => 'application/octet-stream'
];
$upload_pdf_body = file_get_contents('/path/to/pdf_file');
$upload_pdf_request = new Request('POST', 'https://api.pdfrest.com/upload', $upload_pdf_headers, $upload_pdf_body);
$upload_pdf_res = $upload_pdf_client->sendAsync($upload_pdf_request)->wait();
echo $upload_pdf_res->getBody() . PHP_EOL;

$upload_pdf_response_json = json_decode($upload_pdf_res->getBody());

$uploaded_pdf_id = $upload_pdf_response_json->{'files'}[0]->{'id'};

echo "PDF file successfully uploaded with an id of: " . $uploaded_pdf_id . PHP_EOL;


$upload_attachment_client = new Client(['http_errors' => false]);
$upload_attachment_headers = [
  'api-key' => 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx',
  'content-filename' => 'attachment_filename.pdf',
  'Content-Type' => 'application/octet-stream'
];
$upload_attachment_body = file_get_contents('/path/to/attachment_file');
$upload_attachment_request = new Request('POST', 'https://api.pdfrest.com/upload', $upload_attachment_headers, $upload_attachment_body);
$upload_attachment_res = $upload_attachment_client->sendAsync($upload_attachment_request)->wait();
echo $upload_attachment_res->getBody() . PHP_EOL;

$upload_attachment_response_json = json_decode($upload_attachment_res->getBody());

$uploaded_attachment_id = $upload_attachment_response_json->{'files'}[0]->{'id'};

echo "Attachment file successfully uploaded with an id of: " . $uploaded_attachment_id . PHP_EOL;



$add_attachment_client = new Client(['http_errors' => false]);
$add_attachment_headers = [
  'api-key' => 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx',
  'Content-Type' => 'application/json'
];
$add_attachment_body = '{"id":"'.$uploaded_pdf_id.'", "id_to_attach": "'.$uploaded_attachment_id.'"}';
$add_attachment_request = new Request('POST', 'https://api.pdfrest.com/pdf-with-added-attachment', $add_attachment_headers, $add_attachment_body);
$add_attachment_res = $add_attachment_client->sendAsync($add_attachment_request)->wait();
echo $add_attachment_res->getBody() . PHP_EOL;

File: Java/Endpoint Examples/JSON Payload/PDFWithAddedAttachment.java
import io.github.cdimascio.dotenv.Dotenv;
import java.io.File;
import java.io.IOException;
import java.util.concurrent.TimeUnit;
import okhttp3.*;
import org.json.JSONArray;
import org.json.JSONObject;

public class PDFWithAddedAttachment {

  // Specify the path to your file here, or as the first argument when running the program.
  private static final String DEFAULT_FILE_PATH = "/path/to/file.pdf";

  // Specify the path to your file attachment here, or as the second argument when running the
  // program.
  private static final String DEFAULT_ATTACHMENT_PATH = "/path/to/file.xml";

  // Specify your API key here, or in the environment variable PDFREST_API_KEY.
  // You can also put the environment variable in a .env file.
  private static final String DEFAULT_API_KEY = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx";

  public static void main(String[] args) {
    File inputFile, attachmentFile;
    if (args.length > 1) {
      inputFile = new File(args[0]);
      attachmentFile = new File(args[1]);
    } else {
      inputFile = new File(DEFAULT_FILE_PATH);
      attachmentFile = new File(DEFAULT_ATTACHMENT_PATH);
    }

    final Dotenv dotenv = Dotenv.configure().ignoreIfMalformed().ignoreIfMissing().load();

    String uploadFileString = uploadFile(inputFile, "file.pdf");
    JSONObject uploadFileJSON = new JSONObject(uploadFileString);
    if (uploadFileJSON.has("error")) {
      System.out.println("Error during upload: " + uploadFileString);
      return;
    }
    JSONArray fileArray = uploadFileJSON.getJSONArray("files");
    JSONObject fileObject = fileArray.getJSONObject(0);
    String uploadedFileID = fileObject.get("id").toString();

    String uploadAttachmentString = uploadFile(attachmentFile, "file.png");
    JSONObject uploadAttachmentJSON = new JSONObject(uploadAttachmentString);
    if (uploadAttachmentJSON.has("error")) {
      System.out.println("Error during upload: " + uploadAttachmentString);
      return;
    }
    JSONArray attachmentFileArray = uploadAttachmentJSON.getJSONArray("files");
    JSONObject attachmentObject = attachmentFileArray.getJSONObject(0);
    String uploadedAttachmentID = attachmentObject.get("id").toString();

    String JSONString =
        String.format(
            "{\"id\":\"%s\", \"id_to_attach\":\"%s\"}", uploadedFileID, uploadedAttachmentID);

    final RequestBody requestBody =
        RequestBody.create(JSONString, MediaType.parse("application/json"));

    Request request =
        new Request.Builder()
            .header("Api-Key", dotenv.get("PDFREST_API_KEY", DEFAULT_API_KEY))
            .url("https://api.pdfrest.com/pdf-with-added-attachment")
            .post(requestBody)
            .build();
    try {
      OkHttpClient client =
          new OkHttpClient().newBuilder().readTimeout(60, TimeUnit.SECONDS).build();

      Response response = client.newCall(request).execute();
      System.out.println("Processing Result code " + response.code());
      if (response.body() != null) {
        System.out.println(prettyJson(response.body().string()));
      }
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  private static String prettyJson(String json) {
    // https://stackoverflow.com/a/9583835/11996393
    return new JSONObject(json).toString(4);
  }

  // This function is just a copy of the 'Upload.java' file to upload a binary file
  private static String uploadFile(File inputFile, String filename) {

    final Dotenv dotenv = Dotenv.configure().ignoreIfMalformed().ignoreIfMissing().load();

    final RequestBody requestBody =
        RequestBody.create(inputFile, MediaType.parse("application/pdf"));

    Request request =
        new Request.Builder()
            .header("Api-Key", dotenv.get("PDFREST_API_KEY", DEFAULT_API_KEY))
            .header("Content-Filename", filename)
            .url("https://api.pdfrest.com/upload")
            .post(requestBody)
            .build();
    try {
      OkHttpClient client = new OkHttpClient().newBuilder().build();
      Response response = client.newCall(request).execute();
      System.out.println("Upload Result code " + response.code());
      if (response.body() != null) {
        return response.body().string();
      }
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
    return "";
  }
}
